File: AUGUST-2025/07-08-2025/p3363.cs
public class Solution {
    [System.Runtime.CompilerServices.SkipLocalsInit]
    public int MaxCollectedFruits(int[][] fruits) {
        int n=fruits.Length, n1=n-1;

        Span<int> ping=stackalloc int[n], pong=stackalloc int[n];
        
        // start with begin/end positions
        int ans=fruits[0][0] + fruits[^1][^1];
        
        // add from top/right to bottom/right
        int last=fruits[0][^1];
        if(n>2) {
            ping.Clear();
            ping[n1]=last;
            for(int i=1;;) {
                pong.Clear();
                var r=fruits[i];
                ans+=r[i++]; // add diagonal along the way
                for(int j=n-i>i?n-i:i; j<n;) {
                    int oj=j, best=ping[j++];
                    if(oj>0 && ping[oj-1]>best) best=ping[oj-1];
                    if(j<n && ping[j]>best) best=ping[j];
                    last=pong[oj]=best+r[oj];
                }
                if(i>=n1) break;

                ping.Clear();
                r=fruits[i];
                ans+=r[i++]; // add diagonal along the way
                for(int j=n-i>i?n-i:i; j<n;) {
                    int oj=j, best=pong[j++];
                    if (oj>0 && pong[oj-1]>best) best=pong[oj-1];
                    if (j<n && pong[j]>best) best=pong[j];
                    last=ping[oj]=best+r[oj];
                }
                if(i>=n1) break;
            }
        }
        ans+=last;

        // add from bottom/left to bottom/right
        last=fruits[^1][0];
        if(n>2) {
            ping.Clear();
            ping[n1]=last;
            for(int i=1;;) {
                pong.Clear();
                int oi=i++;
                for(int j=n-i>i?n-i:i; j<n;) {
                    int oj=j, best=ping[j++];
                    if(oj>0 && ping[oj-1]>best) best=ping[oj-1];
                    if(j<n && ping[j]>best) best=ping[j];
                    last=pong[oj]=best+fruits[oj][oi];
                }
                if(i>=n1) break;

                ping.Clear();
                oi=i++;
                for(int j=n-i>i?n-i:i; j<n;) {
                    int oj=j, best=pong[j++];
                    if(oj>0 && pong[oj-1]>best) best=pong[oj-1];
                    if(j<n && pong[j]>best) best=pong[j];
                    last=ping[oj]=best+fruits[oj][oi];
                }
                if(i>=n1) break;
            }
        }
        return ans+last;
    }
}

File: AUGUST-2025/07-08-2025/p3363.kt
class Solution {
    fun maxCollectedFruits(fruits: Array<IntArray>): Int {
        val n = fruits.size
        val dp = Array(n) { IntArray(n) { -1 } }
        dp[0][0] = fruits[0][0]
        for (i in 1 until n) {
            dp[i][i] = dp[i - 1][i - 1] + fruits[i][i]
        }
        dp[0][n - 1] = fruits[0][n - 1]
        for (i in 1 until n - 1) {
            for (j in i + 1 until n) {
                dp[i][j] = maxOf(
                    dp[i - 1][j - 1],
                    dp[i - 1][j],
                    if (j + 1 < n) dp[i - 1][j + 1] else -1
                )
                if (dp[i][j] != -1) dp[i][j] += fruits[i][j]
            }
        }
        dp[n - 1][0] = fruits[n - 1][0]
        for (j in 1 until n - 1) {
            for (i in n - 1 downTo j + 1) {
                dp[i][j] = maxOf(
                    if (i + 1 < n) dp[i + 1][j - 1] else -1,
                    dp[i][j - 1],
                    dp[i - 1][j - 1]
                )
                if (dp[i][j] != -1) dp[i][j] += fruits[i][j]
            }
        }
        return dp[n - 1][n - 1] + dp[n - 2][n - 1] + dp[n - 1][n - 2]
    }
}